docs/gc.md

```diff
@@ -141,15 +141,17 @@ Object ages are tracked for each region with a maximum of 24 possible generation
+The *eden* space is a set of regions of age 0, which contain the newest objects allocated. Regions of intermediate age or empty regions serve as survivor regions to receive live objects moved from younger regions. When the eden space is full, a partial GC cycle (PGC) runs to reclaim space, typically by using a *copy forward* operation. In specific cases, a *mark-compact* operation might be used, for example, when there are not enough free survivor regions available. The PGC is a *stop-the-world* (STW) operation, after which new regions are allocated to create a new eden space. A PGC always includes the eden space, but might include older regions. Objects from collectible regions of age N are moved into another region of the same age N or to an empty region that is assigned an age of N. Then, the ages of all regions across the heap are incremented by 1, except for the maximum age 24 regions. Regions of age 24 are included in PGC collection sets in order to defragment them.
 
-Over time, transient objects drop out of the live set and viable survivor regions become scarce. This problem is solved by introducing a global mark phase (GMP) to mark and sweep the entire heap. GMP is a hybrid *incremental-concurrent* GC cycle. Each GMP increment is interleaved with a PGC cycle so that regions that are left sparse after sweeping the heap can be collected, cleared, and recycled during and after the GMP.
+Partial GC cycles reclaim free regions by moving objects out of *eden* and older regions. However, moving objects out of *eden* regions can never reclaim all of the consumed memory because some objects from *eden* regions will survive. The partial GC cycle compensates by moving objects to regions that contain older generations so that newer regions can be defragmented. However, over time the overall amount of defragmented memory decreases and records about object liveness in older regions becomes less accurate. Eventually, the work that is done by partial GC cycles to reclaim memory cannot keep pace with memory allocation. Free regions become so scarce that global *mark* operations (sometimes known as the *Global mark phase* (GMP)) are required to build a new record of object liveness across the heap. This record allows a global *sweep* operation to reclaim memory. The global *sweep* operation that is logically associated with the global *mark* operations, runs in the first partial GC cycle after the *mark* operations complete. Because the GC cycle responsible for the global *mark* operations runs concurrently, it might overlap and interleave with a few partial GC cycles.
```

Comment:
This is not correct 

'The partial GC cycle compensates by moving objects to regions that contain older generations so that newer regions can be defragmented.'

what is reworded variant of my sentence:

'This incomplete reclamation can be compensated through evacuation of older regions through their defragmentation. '

The point is that older (not just Eden) regions become the source of reclaimed memory. For example, we include 5 fragmented older regions, evacuate them, and move objects into 2 new (previously empty regions. So the net effect is that we reclaimed 3 empty regions. 

Also this needs to be corrected:
... over time, he overall amount of fragmented (not defragmented) memory decreases....

Reply:
I also originally stated this, that you omitted: 

`Additionally, a sweep operation is performed to quantify amount of fragmented memory in older regions based on newly built liveness information'

Sweep has dual functionality. First one is that it reclaims memor -, it may create completely empty regions. For those that are not completely freedd, the reclamation of free memory will occur later in PGCs through their defragmentation, as explained above. However, sweep for them has that the second role: it determines how much free (fragmented) memory there is, so that later PGCs know what/when to pick to defragmenent.

It's nice to mention that second role, but it's not mandatory, one reason being that we have plans in future to measure fragmentation within GMP, rather than sweep.

Reply:
I'm still finding your input difficult to understand, but I've had a further pass at trying to get this right. I am still partly troubled that we are introducing a global *sweep* that we don't mention in the table. This *sweep* now does 2 things - measures fragmentation and later reclaims memory? Please take another look at the updated text.

Reply:
Sweep is not a stand alone cycle, it's an operation physically attached to PGC cycle/increment, but logically associated with GMP cycle. So it definitely does not deserve a separate row in the table, but can be mentioned in either PGC or GMP definition, depending if physical or logical association is considered more important. I'm slightly more inclined to mention it as an operation in PGC, since PGC already lists most of operations it does.

Reply:
I think the confusion for me (and potentially users) is that we are talking about a *partial* GC cycle (defined as operating on part of the heap) and then explaining that it does a *global* sweep (defined as operating on the whole heap). I think it helps to explain that the sweep is logically associated with the global mark (phase),  but is triggered(?)/run(?) during a PGC?

Reply:
I think what we have now is fine.

Reply:
yes, table looks reasonable to me know. 

note that beside global sweep, also a local sweep can appear in PGC that operate only on collection set, like compact operation too. so having sweep in PGC in table is definitely reasonable. both sweep operations in VGC are named same though (just 'sweep') so it's not immediately clear which one is whicvh